### PR TITLE
benchmark: change the way of timestamp buffer allocation.

### PR DESCRIPTION
**Pull request by igoropaniuk**
In case if timestamp buffer is allocated in userspace and new register
user memory API is used for its registering in OP-TEE (introduced in
https://github.com/OP-TEE/optee_client/commit/27888d73d156d4862c07effce7c8c2455774768b),
there is no possibility to keep this mapping permanent among different
TEEC_InvokeCommand invocations. All all SHM are automatically unmapped from
OP-TEE VA space after TEEC_InvokeCommand is handled by OP-TEE.

Fixes: https://github.com/OP-TEE/optee_os/issues/1979

Depends on https://github.com/OP-TEE/optee_os/pull/2124
```
Signed-off-by: Igor Opaniuk <igor.opaniuk@linaro.org>
```

### CODE REVIEW

**Comment by jforissier** [comment]
In the commit log:
- The `optee_client` commit is not properly quoted. Should be: '...introduced in optee_client commit 27888d73d156 ("tee_client_api: register user memory")'
- You're not saying what the patch is doing to address the issue. I think you should say the timestamp buffer is now allocated with `thread_rpc_alloc_payload()`.
- Typo: 'All all'.

Other than that:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

**Comment by jforissier** [comment]
In the commit log:
- The `optee_client` commit is not properly quoted. Should be: '...introduced in optee_client commit 27888d73d156 ("tee_client_api: register user memory")'
- You're not saying what the patch is doing to address the issue. I think you should say the timestamp buffer is now allocated with `thread_rpc_alloc_payload()`.
- Typo: 'All all'.

Other than that:
`Acked-by: Jerome Forissier <jerome.forissier@linaro.org>`

File: core/arch/arm/pta/benchmark.c [hunk 1]
```diff
@@ -62,43 +69,61 @@ static TEE_Result register_benchmark_memref(uint32_t type,
...
 
 	/* Send back to the optee linux kernel module */
 	res = rpc_reg_global_buf(OPTEE_MSG_RPC_CMD_BENCH_REG_NEW,
 			virt_to_phys((void *)bench_ts_global),
-			sizeof(struct tee_ts_global) +
-			sizeof(struct tee_ts_cpu_buf) *
-			bench_ts_global->cores);
+			bench_ts_size);
+
+	p[0].value.a = virt_to_phys(bench_ts_global);
```
Comment (etienne-lms) [inline core/arch/arm/pta/benchmark.c]: case 64bit phys address?

Comment (etienne-lms) [reply core/arch/arm/pta/benchmark.c]: seems same issue was already in `get_benchmark_memref()`. Maybe to be fixed is a specific P-R.

File: core/arch/arm/pta/benchmark.c [hunk 2]
```diff
@@ -62,43 +69,61 @@ static TEE_Result register_benchmark_memref(uint32_t type,
...
+	bench_mobj = thread_rpc_alloc_payload(bench_ts_size, &bench_c);
+	if (!bench_mobj) {
+		EMSG(TA_PRINT_PREFIX
+			"can't create mobj for timestamp buffer\n");
+		mutex_unlock(&bench_reg_mu);
+		return TEE_ERROR_OUT_OF_MEMORY;
+	}
+
+	bench_ts_global = (struct tee_ts_global *)
+		mobj_get_va(bench_mobj, 0);
```
Comment (etienne-lms) [inline core/arch/arm/pta/benchmark.c]: minor: fits in a 80 char line.

File: core/arch/arm/pta/benchmark.c [hunk 3]
```diff
@@ -62,43 +69,61 @@ static TEE_Result register_benchmark_memref(uint32_t type,
 {
 	TEE_Result res;
 
-	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_MEMREF_INOUT) ||
+	if ((TEE_PARAM_TYPE_GET(type, 0) != TEE_PARAM_TYPE_VALUE_INOUT) ||
```
Comment (jbech-linaro) [inline core/arch/arm/pta/benchmark.c]: Based on how the function has been changed, it should maybe be renamed from `register_...` to `allocate_...`?

File: core/arch/arm/pta/benchmark.c (hunk 1 above)
Comment (jbech-linaro) [reply core/arch/arm/pta/benchmark.c]: I agree, probably cleaner to deal with that in a separate PR.

Comment (igoropaniuk) [reply core/arch/arm/pta/benchmark.c]: @etienne-lms thanks, will fix

File: core/arch/arm/pta/benchmark.c (hunk 3 above)
Comment (igoropaniuk) [reply core/arch/arm/pta/benchmark.c]: @jbech-linaro will fix

Comment (igoropaniuk) [reply core/arch/arm/pta/benchmark.c]: fixed

**Comment by igoropaniuk** [comment]
Just in case, the build is failing because of `thread_rpc_alloc_dma` and `thread_rpc_free_dma` functions, introduced in https://github.com/OP-TEE/optee_os/pull/2124

File: core/arch/arm/pta/benchmark.c [hunk 4]
```diff
@@ -33,16 +33,23 @@
 #include <mm/core_memprot.h>
 #include <mm/tee_mm.h>
 #include <mm/tee_pager.h>
+#include <mm/mobj.h>
```
Comment (etienne-lms) [inline core/arch/arm/pta/benchmark.c]: minor: move this 2 lines above.

File: core/arch/arm/pta/benchmark.c [hunk 5]
```diff
@@ -33,16 +33,23 @@
...
 #include <string.h>
 #include <string_ext.h>
 #include <stdio.h>
 #include <trace.h>
 
 #define TA_NAME		"benchmark.ta"
+#define TA_PRINT_PREFIX	"Benchmark: "
+
+static volatile struct tee_ts_global *bench_ts_global;
+size_t bench_ts_size;
```
Comment (etienne-lms) [inline core/arch/arm/pta/benchmark.c]: not `static`?

File: core/arch/arm/pta/benchmark.c [hunk 6]
```diff
@@ -57,48 +64,69 @@ static TEE_Result rpc_reg_global_buf(uint64_t type, paddr_t phta, size_t size)
...
 	if (bench_ts_global) {
-		EMSG("Timestamp buffer was already registered\n");
+		EMSG(TA_PRINT_PREFIX
+			"timestamp buffer was already registered\n");
+		mutex_unlock(&bench_reg_mu);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	bench_ts_size = sizeof(struct tee_ts_global) +
+		(p[1].value.a * sizeof(struct tee_ts_cpu_buf));
```
Comment (etienne-lms) [inline core/arch/arm/pta/benchmark.c]: minor: parentheses not needed

File: core/arch/arm/pta/benchmark.c [hunk 7]
```diff
@@ -33,16 +33,23 @@
...
 #define TA_NAME		"benchmark.ta"
+#define TA_PRINT_PREFIX	"Benchmark: "
+
+static volatile struct tee_ts_global *bench_ts_global;
+size_t bench_ts_size;
 
-struct tee_ts_global *bench_ts_global;
 static struct mutex bench_reg_mu = MUTEX_INITIALIZER;
+static struct mobj *bench_mobj;
+static uint64_t bench_c;
```
Comment (etienne-lms) [inline core/arch/arm/pta/benchmark.c]: `bench_cookie` is more friendly.

File: core/arch/arm/pta/benchmark.c [hunk 8]
```diff
@@ -33,16 +33,23 @@
...
 #include <pta_benchmark.h>
 #include <string.h>
 #include <string_ext.h>
 #include <stdio.h>
 #include <trace.h>
 
 #define TA_NAME		"benchmark.ta"
+#define TA_PRINT_PREFIX	"Benchmark: "
+
+static volatile struct tee_ts_global *bench_ts_global;
```
Comment (etienne-lms) [inline core/arch/arm/pta/benchmark.c]: why `volatile`?

File: core/arch/arm/pta/benchmark.c [hunk 9]
```diff
@@ -57,48 +64,69 @@ static TEE_Result rpc_reg_global_buf(uint64_t type, paddr_t phta, size_t size)
...
+			"timestamp buffer was already registered\n");
+		mutex_unlock(&bench_reg_mu);
+		return TEE_ERROR_BAD_STATE;
+	}
+
+	bench_ts_size = sizeof(struct tee_ts_global) +
+		(p[1].value.a * sizeof(struct tee_ts_cpu_buf));
+	if (!bench_ts_size) {
+		EMSG(TA_PRINT_PREFIX
+			"timestamp buffer size isn't specified\n");
```
Comment (etienne-lms) [inline core/arch/arm/pta/benchmark.c]: The issue looks more like `invalid timestamp buffer size found`.

minor: no need to add ending `\n` in `EMSG()` macro and friends.

**Comment by igoropaniuk** [comment]
@etienne-lms addressed all your comments

**Comment by igoropaniuk** [comment]
@jbech-linaro @jforissier, although I've cross-checked that all your comments were addressed, could you please confirm that I can still keep you A-b/R-b tags.

Thanks!

**Comment by jforissier** [comment]
@igoropaniuk LGTM

**Comment by etienne-lms** [comment]
`Acked-by: Etienne Carriere <etienne.carriere@linaro.org>`

**Comment by jbech-linaro** [comment]
Feel free to use my R-B after the doing recent changes.

**Comment by igoropaniuk** [comment]
Attached tags, rebased on the latest master and ready for merge. Thanks

**Pull request merged**